lib/ContentTypeParser.js

```diff
@@ -1,67 +1,181 @@
-    result
-      .then(body => done(null, body))
-      .catch(done)
+  var parser = this.cache.get[contentType] || this.getParser(contentType)
+  if (parser === undefined) {
+    reply.code(415).send(new Error('Unsupported Media Type: ' + contentType))
+  } else {
+    if (parser.asString === true || parser.asBuffer === true) {
+      rawBody(
+        request,
+        reply,
+        reply.context._parserOptions,
+        parser,
+        handler
+      )
+    } else {
+      var result = parser(request.raw, done)
+      if (result && typeof result.then === 'function') {
+        result.then(body => done(null, body), done)
+      }
+    }
+  }
+
+  function done (error, body) {
+    if (error) {
+      reply.send(error)
+    } else {
+      request.body = body
+      handler(reply)
+    }
+  }
+}
+
+function rawBody (request, reply, options, parser, handler) {
+  var asString = parser.asString
+  var limit = options.limit
+  var contentLength = request.headers['content-length'] === undefined
+    ? NaN
+    : Number.parseInt(request.headers['content-length'], 10)
+
+  if (contentLength > limit) {
+    reply.code(413).send(new Error('Request body is too large'))
+    return
+  }
+
+  var receivedLength = 0
+  var body = asString === true ? '' : []
+  var req = request.raw
+
+  req.on('data', onData)
+  req.on('end', onEnd)
+  req.on('error', onEnd)
+
+  function onData (chunk) {
+    receivedLength += chunk.length
+
+    if (receivedLength > limit) {
+      req.removeListener('data', onData)
+      req.removeListener('end', onEnd)
+      req.removeListener('error', onEnd)
+      reply.code(413).send(new Error('Request body is too large'))
+      return
+    }
+
+    if (asString === true) {
+      body += chunk.toString()
+    } else {
+      body.push(chunk)
+    }
+  }
+
+  function onEnd (err) {
+    req.removeListener('data', onData)
+    req.removeListener('end', onEnd)
+    req.removeListener('error', onEnd)
+
+    if (err !== undefined) {
+      reply.code(400).send(err)
+      return
+    }
+
+    if (!Number.isNaN(contentLength) && receivedLength !== contentLength) {
+      reply.code(400).send(new Error('Request body size did not match Content-Length'))
+      return
+    }
+
+    if (receivedLength === 0) {
+      reply.code(400).send(new Error('Unexpected end of body input'))
+      return
+    }
+
+    if (asString === false) {
+      body = Buffer.concat(body)
+    }
+
+    var result = parser(request, reply, body, done)
+    if (result && typeof result.then === 'function') {
+      result.then(body => done(null, body), done)
+    }
   }
 
   function done (error, body) {
     if (error) {
-      return reply.send(error)
+      reply.send(error)
+    } else {
+      request.body = body
+      handler(reply)
     }
+  }
+}
 
-    request.body = body
-    handler(reply)
+function defaultJsonParser (request, reply, body, done) {
+  try {
+    var json = JSON.parse(body)
+    done(null, json)
```

Comment:
This should be called outside of the try-catch (so that you don't catch errors in `done` and then call the `done` function a second time).